### ISSUE TITLE
Multiprocessing of search gives no [query] registered for [sltr] for ~20% of queries

### ISSUE THREAD

**Issue by lasagar** [o19s/elasticsearch-learning-to-rank#301]
Multiprocessing of search gives no [query] registered for [sltr] for ~20% of queries everytime i execute it. This is the part of code from collect_features.py where am using multiprocessing. Any suggestion to mitigate this issue and process the queries in parallel.

def log_features(es, judgments_dict, search_index):
    print("Inside log features")
    global start_time
    start_time = time.time()
    print(start_time)

    queries_list =[]
    temp =dict()

    for qid, judgments in judgments_dict.items():
        temp[qid] = judgments
        queries_list.append(temp)
        temp = dict()

    # Run this with a pool of 5 agents having a chunksize of 3 until finished
    agents = 3
    keywords_chunks = 2
    with Pool(processes=agents) as pool:
        result = pool.map(get_scores, queries_list, keywords_chunks)

    print("--- %s seconds ---" % (time.time() - start_time))

    return result

def get_scores(judgments):
    es = elastic_connection()
    search_index = INDEX_NAME

    keywords = list(judgments.values())[0][0].keywords
    doc_ids = [judgment.docId for judgment in list(judgments.values())[0]]
    logQuery["query"]["bool"]["filter"][0]["terms"]["url.keyword"] = doc_ids
    logQuery["query"]["bool"]["should"][0]["sltr"]["params"]["keywords"] = keywords
    logQuery["query"]["bool"]["should"][0]["sltr"]["featureset"] = FEATURE_SET_NAME
    Logger.logger.info("POST")
    Logger.logger.info(json.dumps(logQuery, indent=2))
    # print(logQuery)

    try:
        res = es.search(index=search_index, body=logQuery)
        # print("Success log")
        # Add feature back to each judgment
        features_per_doc = {}
        for doc in res['hits']['hits']:
            # docId = doc['_id']
            docId = doc['_source']['url']
            features = doc['fields']['_ltrlog'][0]['main']
            features_per_doc[docId] = feature_dict_to_list(features)

        # Append features from ES back to ranklib judgment list
        for k, judgment in judgments.items():
            for i in judgment:
                try:
                    features = features_per_doc[i.docId]  # If KeyError, then we have a judgment but no URL in index
                    i.features = features
                except KeyError:
                    Logger.logger.info("Missing Document %s" % i.docId)
                    i.features = [0.0] * int(FEATURES_COUNT)
    except Exception as e:
        Logger.logger.info("Search Exception is for query {} with exception {}".format(keywords, e))
        for k, judgment in judgments.items():
            for i in judgment:
                i.features = [0.0] * int(FEATURES_COUNT)
            pass

    return judgments

**Comment by softwaredoug** [comment]
Hi @lasagar - can you confirm ltr plugin is installed on every node in your cluster? You can use the [`_cat/plugins` API](https://www.elastic.co/guide/en/elasticsearch/reference/current/cat-plugins.html)

**Comment by lasagar** [comment]
@softwaredoug Yes , i see them in all 4 nodes.

**Comment by softwaredoug** [comment]
You are sure you have 4 nodes? The Elasticsearch default is 5 shards.

I would recommend forking this repo, and creating a demo that recreates the issue using just the demo code. As it's hard to know if the issue is with your ES setup or with the plugin itself...

**Comment by lasagar** [comment]
The issue is only with multiprocessing, otherwise its working fine

**Comment by softwaredoug** [comment]
Yeah I think we need a way to recreate the issue without it involving your setup to debug the issue. Such as in a standalone environment with instructions. we could access and recreate locally.

We are happy to provide support of course of your LTR setup, but that would be a consulting relationship

**Comment by lasagar** [comment]
Sure, will get back
